### PR TITLE
add live_file_input component & test

### PR DESCRIPTION
**Pull request by harmon25**
Initial PR for #298 

Test does not do much but verify it renders the expected phoenix generated content. Should we do anything else?

Happy to iterate a bit on this, feedback is very appreciated!

### CODE REVIEW

File: lib/surface/components/live_file_input.ex
Comment (Malian) [inline lib/surface/components/live_file_input.ex:21]: Usually we do not give a default value to the `class` prop. It is nil by default.

File: lib/surface/components/live_file_input.ex
Comment (Malian) [inline lib/surface/components/live_file_input.ex:26]: As we should have the most reusable component across project we usually provide a way to define a `default_class` to Surface component through the config.

Take a look at the other components to know how to do that. For example, the [ColorInput](https://github.com/surface-ui/surface/blob/master/lib/surface/components/form/color_input.ex) component

File: test/components/live_file_input_test.exs
Comment (Malian) [inline test/components/live_file_input_test.exs:18]:
```suggestion
        <LiveFileInput upload={{ @uploads.test }} />
```

File: lib/surface/components/live_file_input.ex
Comment (Malian) [inline lib/surface/components/live_file_input.ex:10]:
```suggestion
      socket = allow_upload(socket, :avatar, accept: ~w(.json), max_entries: 1)
```

Just to be consistent with the doc ;)

**Review by Malian — CHANGES_REQUESTED** [review]
@harmon25 Thanks for the pull request.

I let a few comments regarding the code and below you will find more general comments:
- I believe we should add a `name` prop to be consistent with all input we defined in Surface. See the `component/form/input.ex` file.
- Regarding the test, it's ok like this. We could just add a test to test that if a class is passed through the prop, it's correctly rendered, and a test that if no class is passed through the prop and if a default class is set in the config, it's also correctly rendered.

WDYT?

**Comment by harmon25** [comment]
Thanks @Malian!
I made most of the changes, will add more tests!

> I believe we should add a name prop to be consistent with all input we defined in Surface. See the component/form/input.ex file.

`live_file_input` being a liveview helper - it does not behave like a phoenix form helper. The name is being assigned based on the atom passed into `allow_upload` and is ignored if passed in as an opt. It does appear to accept an ID opt- which otherwise is being set by LiveView to the same value as `data-phx-upload-ref`.  (not sure if this will break the live_upload, need to test more)

Here is a sample of the rendered output: 

```html
<input accept=\".json\" class=\"test_class disabled_test\" data-phx-active-refs=\"\" data-phx-done-refs=\"\" data-phx-preflighted-refs=\"\" data-phx-update=\"ignore\" data-phx-upload-ref=\"phx-Fm8ZSnh8_rSc_RWH\" data-test=\"test-data\" id=\"phx-Fm8ZSnh8_rSc_RWH\" name=\"test\" phx-hook=\"Phoenix.LiveFileUpload\" type=\"file\"></input>\n
```

I am going to test it out a bit more in an app, and try writing a test using [render_upload](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveViewTest.html#render_upload/3) among the other tests you suggested.

**Comment by harmon25** [comment]
on a closer look at the `live_file_input/2` docs, passing ID should be OK.

> By default, the file input ID is the upload ref...

File: lib/surface/components/live_file_input.ex
Comment (Malian) [inline lib/surface/components/live_file_input.ex:25]: This will not work anymore with PhoenixLiveView `0.15.2`, see the following commit: https://github.com/phoenixframework/phoenix_live_view/commit/c263db32f446e57275f62f507f4995fef59c1975 we could simply remove it.

File: lib/surface/components/live_file_input.ex
Comment (Malian) [inline lib/surface/components/live_file_input.ex:33]: After a close look at the `live_file_input/2` function, it seems that this method does not accept any options and these options are simply forwarded to the underlying `content_tag` function that generates the `input`. I suggest to refrasing to something more general:

```suggestion 
  @doc "Keyword list with options to be passed down to `live_file_input/2`"
```

File: lib/surface/components/live_file_input.ex
Comment (Malian) [inline lib/surface/components/live_file_input.ex:37]:
```suggestion
    attr_opts = props_to_attr_opts(assigns, [class: get_config(:default_class)])
```

File: test/components/live_file_input_test.exs
Comment (Malian) [inline test/components/live_file_input_test.exs:19]:
```suggestion
        <LiveFileInput upload={{ @uploads.avatar }} class={{ "test_class", disabled_test: true  }} opts={{ "data-test": "test-data", name: "a name?" }} />
```

File: test/components/live_file_input_test.exs
Comment (Malian) [inline test/components/live_file_input_test.exs:37]:
```suggestion
        <LiveFileInput upload={{ @uploads.avatar }} />
```

File: test/components/live_file_input_test.exs
Comment (Malian) [inline test/components/live_file_input_test.exs:101]:
```suggestion
```

No need due to remove of `id`

File: test/components/live_file_input_test.exs
Comment (Malian) [inline test/components/live_file_input_test.exs:2]: I added a test to use the `:default_class` of the `LiveFileInput` component. That test requires that async must be set to `false`
```suggestion
  use Surface.ConnCase
```

File: test/components/live_file_input_test.exs
Comment (Malian) [inline test/components/live_file_input_test.exs:89]:
```suggestion
  test "correctly renders live_file_input/2 with `:default_class` config" do
    using_config LiveFileInput, default_class: "default_class" do
      html =
        render_surface do
          ~H"""
          <LiveFileInputWithoutProps />
          """
        end

      assert html =~ ~r/class="default_class"/
    end
  end
```

File: test/components/live_file_input_test.exs
Comment (Malian) [inline test/components/live_file_input_test.exs:75]:
```suggestion
```

**Review by Malian — CHANGES_REQUESTED** [review]
@harmon25 Thank for the changes, but there are still some small changes to be made and we will merge!

After a close look at the last versions of Live View, the `id` prop will stop work with Phoenix Live View `0.15.2`. After several tests, it's better to remove it right now. Most of my comments are related to this change. It was your comment about the name that tipped me off :+1: 

I also update a test to check if the `default_class` defined in the configuration is correctly used if provided.

I'll update the documentation in #311 to use the `data :avatar, :upload, accept: ...` new type. Thanks again for this suggestion :+1: 

We are close merging this request! Thanks again for your work!

File: lib/surface/components/live_file_input.ex
Comment (harmon25) [reply lib/surface/components/live_file_input.ex:25]: good catch! 
I'll remove it.

**Comment by miguel-s** [comment]
Thanks for the PR @harmon25 and awesome reviews @Malian!

How does `live_file_input` behave with regular LiveView events like `phx-click` or `phx-blur`? If they support passing in events we should, too.

Example of how we handle events in the regular `file_input` https://github.com/surface-ui/surface/blob/d2f68bd7a6b4ebc5b0ef00b7cdf06f3cce377352/lib/surface/components/form/file_input.ex#L30

**Comment by harmon25** [comment]
Thank you both for helping support Surface! 
I am not sure if that would interfere with the hook that is being used:
https://github.com/phoenixframework/phoenix_live_view/blob/v0.15.4/lib/phoenix_live_view/helpers.ex#L446

**Comment by Malian** [comment]
> How does `live_file_input` behave with regular LiveView events like `phx-click` or `phx-blur`? If they support passing in events we should, too.

@miguel-s   Good catch. Passing events through the second argument of `live_file_input` works perfectly fine, so indeed we should add them as props.

@harmon25 it should be simple as adding the following codes to the component:

https://github.com/surface-ui/surface/blob/d2f68bd7a6b4ebc5b0ef00b7cdf06f3cce377352/lib/surface/components/form/file_input.ex#L30

https://github.com/surface-ui/surface/blob/d2f68bd7a6b4ebc5b0ef00b7cdf06f3cce377352/lib/surface/components/form/input.ex#L32-L45

**Comment by harmon25** [comment]
Added the event props!

**Review by harmon25 — COMMENTED** [review]
this should be ready!

**Review by Malian — APPROVED** [review]
@harmon25 thanks! It should be ready. I am waiting for a decision about #298 before merging it to be sure we do not miss something but normally it's ok :+1:

**Review by miguel-s — APPROVED** [review]

**Comment by miguel-s** [comment]
Looking good to me! Thanks @harmon25 ❤️

**Comment by msaraiva** [comment]
@harmon25 thanks! ❤️

**Pull request merged**
